topology/probes/lldp/lldp.go

```diff
@@ -0,0 +1,424 @@
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ *
+ */
+
+package lldp
+
+import (
+	"fmt"
+	"io"
+	"net"
+	"sync"
+	"sync/atomic"
+	"syscall"
+	"time"
+	"unsafe"
+
+	"github.com/google/gopacket"
+	"github.com/google/gopacket/layers"
+	"github.com/safchain/ethtool"
+	"github.com/skydive-project/skydive/common"
+	"github.com/skydive-project/skydive/flow"
+	"github.com/skydive-project/skydive/flow/probes"
+	"github.com/skydive-project/skydive/flow/probes/afpacket"
+	"github.com/skydive-project/skydive/logging"
+	"github.com/skydive-project/skydive/topology"
+	"github.com/skydive-project/skydive/topology/graph"
+	"golang.org/x/net/bpf"
+	"golang.org/x/sys/unix"
+)
+
+/*
+#include <linux/if_packet.h>  // packet_mreq
+#include <linux/if_ether.h>  // ETH_ALEN
+*/
+import "C"
+
+// Capture 8192 bytes so that we have the full Ethernet frame
+const lldpSnapLen = 8192
+
+// Probe describes the probe that is in charge of listening for
+// LLDP packets on interfaces and create the corresponding chassis and port nodes
+type Probe struct {
+	sync.RWMutex
+	graph.DefaultGraphListener
+	g             *graph.Graph
+	hostNode      *graph.Node
+	interfaceMap  map[string]bool
+	state         int64
+	wg            sync.WaitGroup
+	autoDiscovery bool
+}
+
+type ifreq struct {
+	ifrName   [ethtool.IFNAMSIZ]byte
+	ifrHwaddr syscall.RawSockaddr
+}
+
+func addMulticastAddr(intf string, addr string) error {
+	fd, err := syscall.Socket(syscall.AF_INET, syscall.SOCK_DGRAM, 0)
+	if err != nil {
+		return err
+	}
+	defer syscall.Close(fd)
+
+	var name [ethtool.IFNAMSIZ]byte
+	copy(name[:], []byte(intf))
+
+	mac, _ := net.ParseMAC(addr)
+
+	var sockaddr syscall.RawSockaddr
+	// RawSockaddr.Data is []int8 on amd64 and []uint8 on ppc64le
+	switch data := interface{}(sockaddr.Data).(type) {
+	case [14]int8:
+		for i, n := range mac {
+			data[i] = int8(n)
+		}
+	case [14]uint8:
+		for i, n := range mac {
+			data[i] = uint8(n)
+		}
+	}
+
+	ifr := &ifreq{
+		ifrName:   name,
+		ifrHwaddr: sockaddr,
+	}
+
+	_, _, ep := unix.Syscall(unix.SYS_IOCTL, uintptr(fd),
+		unix.SIOCADDMULTI, uintptr(unsafe.Pointer(ifr)))
+
+	if ep != 0 {
+		return syscall.Errno(ep)
+	}
+	return nil
+}
+
+func (p *Probe) startCapture(ifName, mac string, n *graph.Node) error {
+	var err error
+	var handle *probes.AFPacketHandle
+
+	p.interfaceMap[ifName] = true
+
+	headerSize := uint32(lldpSnapLen)
+	fnc := func() error {
+		handle, err = probes.NewAFPacketHandle(ifName, int32(headerSize))
+		if err != nil {
+			return fmt.Errorf("Error while opening device %s: %s", ifName, err)
+		}
+		return nil
+	}
+
+	if err = common.Retry(fnc, 2, 100*time.Millisecond); err != nil {
+		return fmt.Errorf("Failed to capture packets on %s: %s", ifName, err)
```

Comment:
nit: I think will have twice almost the same message, concatenation of L127 + this one. I would just do return err

Reply:
Removed common.Retry